File: extensions/interactions/SetInput/SetInput.py
# coding: utf-8
#
# Copyright 2014 The Oppia Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, softwar
# distributed under the License is distributed on an "AS-IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Python configuration for SetInput interaction."""

from __future__ import annotations

from extensions.interactions import base

from typing import List, Optional

MYPY = False
if MYPY:  # pragma: no cover
    from extensions import domain


class SetInput(base.BaseInteraction):
    """Interaction for input of an unordered set of strings."""

    name: str = 'Set Input'
    description: str = 'Allows learners to enter an unordered set of strings.'
    display_mode: str = base.DISPLAY_MODE_INLINE
    _dependency_ids: List[str] = []
    answer_type: str = 'SetOfUnicodeString'
    instructions: Optional[str] = None
    narrow_instructions: Optional[str] = None
    needs_summary: bool = False
    can_have_solution: bool = True
    show_generic_submit_button: bool = True

    # NB: There used to be a UnicodeString-typed parameter here called
    # 'element_type'. This has since been removed.
    _customization_arg_specs: List[domain.CustomizationArgSpecsDict] = [{
        'name': 'buttonText',
        'description': 'Label for the \'Add Item\' button',
        'schema': {
            'type': 'custom',
            'obj_type': 'SubtitledUnicode'
        },
        'default_value': {
            'content_id': None,
            'unicode_str': 'Add Item'
        }
    }]

    _answer_visualization_specs: List[base.AnswerVisualizationSpecsDict] = [{
        # Table with answer counts for top N answers.
        'id': 'FrequencyTable',
        'options': {
            'column_headers': ['Answer', 'Count'],
            'title': 'Top 10 answers',
        },
        'calculation_id': 'Top10AnswerFrequencies',
        'addressed_info_is_supported': True,
    }, {
        # Table with most commonly submitted elements of set.
        'id': 'FrequencyTable',
        'options': {
            'column_headers': ['Element', 'Count'],
            'title': 'Commonly submitted elements',
        },
        'calculation_id': 'FrequencyCommonlySubmittedElements',
        # Since individual answer elements are not generally intended to be
        # used as a single response to SetInput interactions, we omit the
        # addressed column entirely.
        'addressed_info_is_supported': False,
    }]

File: extensions/interactions/SetInput/SetInput.js
// Copyright 2014 The Oppia Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS-IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * Directive for the SetInput interaction.
 *
 * IMPORTANT NOTE: The naming convention for customization args that are passed
 * into the directive is: the name of the parameter, followed by 'With',
 * followed by the name of the arg.
 */

oppia.directive('oppiaInteractiveSetInput', [function() {
  return {
    restrict: 'E',
    scope: {
      onSubmit: '&'
    },
    templateUrl: 'interaction/SetInput',
    controller: [
      '$scope', '$attrs', '$translate', 'setInputRulesService',
      function($scope, $attrs, $translate, setInputRulesService) {
        $scope.schema = {
          type: 'list',
          items: {
            type: 'unicode'
          },
          ui_config: {
            // TODO(mili): Translate this in the HTML.
            add_element_text: $translate.instant(
              'I18N_INTERACTIONS_SET_INPUT_ADD_ITEM')
          }
        };

        // Adds an input field by default
        $scope.answer = [''];

        var hasDuplicates = function(answer) {
          for (var i = 0; i < answer.length; i++) {
            for (var j = 0; j < i; j++) {
              if (angular.equals(answer[i], answer[j], true)) {
                return true;
              }
            }
          }
          return false;
        };

        $scope.submitAnswer = function(answer) {
          if (hasDuplicates(answer)) {
            $scope.errorMessage = (
              'I18N_INTERACTIONS_SET_INPUT_DUPLICATES_ERROR');
          } else {
            $scope.errorMessage = '';
            $scope.onSubmit({
              answer: answer,
              rulesService: setInputRulesService
            });
          }
        };
      }]
  };
}]);

oppia.directive('oppiaResponseSetInput', [
  'oppiaHtmlEscaper', function(oppiaHtmlEscaper) {
    return {
      restrict: 'E',
      scope: {},
      templateUrl: 'response/SetInput',
      controller: ['$scope', '$attrs', function($scope, $attrs) {
        $scope.answer = oppiaHtmlEscaper.escapedJsonToObj($attrs.answer);
      }]
    };
  }
]);

oppia.directive('oppiaShortResponseSetInput', [
  'oppiaHtmlEscaper', function(oppiaHtmlEscaper) {
    return {
      restrict: 'E',
      scope: {},
      templateUrl: 'shortResponse/SetInput',
      controller: ['$scope', '$attrs', function($scope, $attrs) {
        var _answer = oppiaHtmlEscaper.escapedJsonToObj($attrs.answer);
        $scope.displayedAnswer = (
          _answer.length > 0 ? _answer.join(', ') :
          'I18N_INTERACTIONS_SET_INPUT_NO_ANSWER');
      }]
    };
  }
]);

oppia.factory('setInputRulesService', [function() {
  return {
    Equals: function(answer, inputs) {
      return answer.length === inputs.x.length && inputs.x.every(function(val) {
        return answer.indexOf(val) >= 0;
      });
    },
    IsSubsetOf: function(answer, inputs) {
      return answer.length < inputs.x.length && answer.every(function(val) {
        return inputs.x.indexOf(val) >= 0;
      });
    },
    IsSupersetOf: function(answer, inputs) {
      return answer.length > inputs.x.length && inputs.x.every(function(val) {
        return answer.indexOf(val) >= 0;
      });
    },
    HasElementsIn: function(answer, inputs) {
      return inputs.x.some(function(val) {
        return answer.indexOf(val) >= 0;
      });
    },
    HasElementsNotIn: function(answer, inputs) {
      return answer.some(function(val) {
        return inputs.x.indexOf(val) === -1;
      });
    },
    OmitsElementsIn: function(answer, inputs) {
      return inputs.x.some(function(val) {
        return answer.indexOf(val) === -1;
      });
    },
    IsDisjointFrom: function(answer, inputs) {
      return inputs.x.every(function(val) {
        return answer.indexOf(val) === -1;
      });
    }
  };
}]);
